win32ss/user/user32/controls/ghost.c

```diff
@@ -0,0 +1,570 @@
+}
+
+static VOID
+IntWhiten32BppBitmap(HBITMAP hbm)
+{
+    BITMAP bm;
+    DWORD i, *pdw;
+
+    GetObject(hbm, sizeof(bm), &bm);
+
+    if (bm.bmBitsPixel != 32 || !bm.bmBits)
+        return;
+
+    pdw = (DWORD *)bm.bmBits;
+    for (i = 0; i < bm.bmWidth * bm.bmHeight; ++i)
+    {
+        *pdw = *pdw | 0x00C0C0C0;   // ARGB
+        ++pdw;
+    }
+}
+
+/****************************************************************************/
+
+typedef struct GHOST_DATA
+{
+    HWND hwndTarget;
+    HBITMAP hbm32bpp;
+} GHOST_DATA;
+
+static GHOST_DATA *
+Ghost_GetData(HWND hwnd)
+{
+    if (IsWindowUnicode(hwnd))
+        return (GHOST_DATA *)GetWindowLongPtrW(hwnd, GWLP_USERDATA);
+    else
+        return (GHOST_DATA *)GetWindowLongPtrA(hwnd, GWLP_USERDATA);
+}
+
+static HWND
+Ghost_GetTarget(HWND hwnd)
+{
+    GHOST_DATA *pData = Ghost_GetData(hwnd);
+    if (!pData)
+        return NULL;
+    return pData->hwndTarget;
+}
+
+static BOOL
+Ghost_OnCreate(HWND hwnd, CREATESTRUCTW *lpcs)
+{
+    HBITMAP hbm32bpp;
+    HWND hwndTarget, hwndPrev;
+    GHOST_DATA *pData;
+    RECT rc;
+    DWORD style, exstyle;
+    CHAR szTextA[128];
+    WCHAR szTextW[128];
+
+#ifdef __REACTOS__
+    PWND pWnd = ValidateHwnd(hwnd);
+    if (pWnd)
+    {
+        if (!pWnd->fnid)
+        {
+            NtUserSetWindowFNID(hwnd, FNID_GHOST);
+        }
+        else if (pWnd->fnid != FNID_GHOST)
+        {
+             ERR("Wrong window class for Ghost! fnId 0x%x\n", pWnd->fnid);
+             return 0;
+        }
+    }
+#endif
+
+    // get the target
+    hwndTarget = (HWND)lpcs->lpCreateParams;
+    if (!hwndTarget || !IsWindowVisible(hwndTarget) || IsIconic(hwndTarget) ||
+        GetParent(hwndTarget))
+    {
+        return FALSE;
+    }
+
+    // create data
+    pData = (GHOST_DATA *)HeapAlloc(GetProcessHeap(), HEAP_ZERO_MEMORY, sizeof(GHOST_DATA));
+    if (!pData)
+        return FALSE;
+
+    // get window image
+    GetWindowRect(hwndTarget, &rc);
+    hbm32bpp = IntGetWindowBitmap(hwndTarget, rc.right - rc.left, rc.bottom - rc.top);
+    if (!hbm32bpp)
+    {
+        HeapFree(GetProcessHeap(), 0, pData);
+        return FALSE;
+    }
+    // whiten
+    IntWhiten32BppBitmap(hbm32bpp);
+
+    // get style and text
+    if (IsWindowUnicode(hwnd))
+    {
+        style = GetWindowLongPtrW(hwndTarget, GWL_STYLE);
+        exstyle = GetWindowLongPtrW(hwndTarget, GWL_EXSTYLE);
+        GetWindowTextW(hwndTarget, szTextW, ARRAYSIZE(szTextW) - 32);
+    }
+    else
+    {
+        style = GetWindowLongPtrA(hwndTarget, GWL_STYLE);
+        exstyle = GetWindowLongPtrA(hwndTarget, GWL_EXSTYLE);
+        GetWindowTextA(hwndTarget, szTextA, ARRAYSIZE(szTextW) - 32);
+    }
+
+    style &= ~(WS_HSCROLL | WS_VSCROLL | WS_VISIBLE);
+
+    // set style and text
+    if (IsWindowUnicode(hwnd))
+    {
+        SetWindowLongPtrW(hwnd, GWL_STYLE, style);
+        SetWindowLongPtrW(hwnd, GWL_EXSTYLE, exstyle);
+
+        lstrcatW(szTextW, L" (Not Responding)");
```

Comment:
This is a buffer overflow waiting to happen. Use strsafe functions and/or allocate an appropriately-sized buffer please.

Reply:
Used <strsafe.h>.